fees/incrementor.go

```diff
@@ -0,0 +1,321 @@
+/* Mysterium network payment library.
+ *
+ * Copyright (C) 2021 BlockDev AG
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package fees
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+// GasPriceIncremenetor exposes a way automatically increment gas fees
+// for transactions in a preconfigured manner.
+type GasPriceIncremenetor struct {
+	pullInterval time.Duration
+
+	storage Storage
+	bc      MultichainClient
+	sign    SignatureFunc
+	logFn   LogFunc
+
+	syncer *syncer
+
+	stop chan struct{}
+	once sync.Once
+}
+
+// Storage is given to the Incremeter to be used to
+// insert, update or get transactions.
+type Storage interface {
+	// UpsertIncrementorTransaction is called to upsert a transaction.
+	// It either inserts a new entry or updates existing entries.
+	UpsertIncrementorTransaction(tx Transaction) error
+
+	// GetIncrementorTransactionsToCheck returns all transaction that need to rechecked.
+	GetIncrementorTransactionsToCheck() (tx []Transaction, err error)
+}
+
+// Client handles calls to BC.
+type Client interface {
+	TransactionReceipt(ctx context.Context, hash common.Hash) (*types.Receipt, error)
+	SendTransaction(ctx context.Context, tx *types.Transaction) error
+}
+
+// SignatureFunc is used to sign transactions when resubmitting them.
+type SignatureFunc func(tx *types.Transaction, chainID int64) (*types.Transaction, error)
+
+// LogFunc can be attacheched to Incrementer to enable logging.
+type LogFunc func(Transaction, error)
+
+// MultichainClient is a map of Client for each chain to process.
+type MultichainClient map[int64]Client
```

Comment:
this package already exports a multichain client, why not use that instead?

Reply:
The interface doesn't have the Transfer method. I guess I can see if that can be added 🤷🏻